pkg/k8s/service_cache.go

```diff
@@ -633,3 +690,65 @@ func (s *ServiceCache) DebugStatus() string {
 	s.mutex.RUnlock()
 	return str
 }
+
+// Implementation of subscriber.Node
+
+func (s *ServiceCache) OnAddNode(node *core_v1.Node, swg *lock.StoppableWaitGroup) error {
+	s.updateSelfNodeLabels(node.GetLabels(), swg)
+
+	return nil
+}
+
+func (s *ServiceCache) OnUpdateNode(oldNode, newNode *core_v1.Node,
+	swg *lock.StoppableWaitGroup) error {
+
+	s.updateSelfNodeLabels(newNode.GetLabels(), swg)
+
+	return nil
+}
+
+func (s *ServiceCache) OnDeleteNode(node *core_v1.Node,
+	swg *lock.StoppableWaitGroup) error {
+
+	return nil
+}
+
+func (s *ServiceCache) updateSelfNodeLabels(labels map[string]string,
+	swg *lock.StoppableWaitGroup) {
+
+	if !option.Config.EnableServiceTopology {
+		return
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	zone, ok := labels[LabelTopologyZone]
+	if !ok && s.selfNodeZoneLabel == "" {
```

Comment:
```suggestion
	if !ok || s.selfNodeZoneLabel == "" {
```
```suggestion
	if !ok && s.selfNodeZoneLabel == "" {
```

Reply:
Did you mean to check if the `zone` is empty instead of `s.selfNodeZoneLabel`?

Reply:
This meant to make it return early when the zone in the given labels is not found, and previously the self node label zone was not set. I can merge it with the next check.